File: src/panic.rs
use core::panic::PanicInfo;
use core::fmt::Write;

pub fn panic_handler(info: &PanicInfo) -> ! {
    // Write panic info to serial (always available, even if VGA is broken)
    let mut serial = crate::drivers::serial::SerialPort;
    let _ = writeln!(serial, "KERNEL PANIC: {}", info);

    loop {
        unsafe { core::arch::asm!("hlt") };
    }
}

File: src/panic.cpp
#include "panic.hpp"
#include "drivers/serial.hpp"
#include "drivers/vga.hpp"

void panic_handler(const PanicInfo& info) {
    // Disable interrupts
    __asm__ volatile("cli");
    
    // Write to serial
    Drivers::Serial::write_string("KERNEL PANIC: ");
    Drivers::Serial::write_string(info.message);
    Drivers::Serial::write_string("\n");
    
    // Write to VGA
    VGA::println("KERNEL PANIC!");
    VGA::print("Error: ");
    VGA::println(info.message);
    
    // Halt
    while (true) {
        __builtin_ia32_hlt();
    }
}